OutSystems/outsystems-maps
# RGRIDT-858: [Maps] Shapes - Polygon block

This PR is for RGRIDT-858: [Maps] Shapes - Polygon block


### What was happening
* We needed to create a Polygon Shape block

### What was done
* Adapted the previously added Polyline into a new AbstractProviderShape which has all the logic that can be used by the remaining shapes (polygone and polyline).
* Divided the GoogleShapeConfig into different files:
  * **BasicShapeConfig** -> related to the shared configs and optionalConfigurations such as strokeColor, strokeOpacity and strokeWeight
  * **FilledShapeConfig** -> related to the optionalConfigurations such as fillColor and fillOpacity (extends all the BasicShapeConfig and is used by the Polygon)
* Cleaned up some code that was required on my last PR (RGRIDT-858)


### Test Steps
<html>
<body>


Shape Test Case | Steps 
-- | --
Polygon Ordered Locations | Click on the Button "Polygon"<br>Set the Locations textarea with ['Areeiro Lisbon, PT', 'Madre Deus Lisbon, PT', 'Alfama Lisbon, PT', 'Jardim da Estrela, Lisbon, PT']<br>**Expected: The Polygon should have 4 edges start in Areeiro, cross Madre Deus, then Alfama and finally in Jardim da Estrela which at the end joins with Areeiro's edge again.**
Platform “if” false Polygon | Click on the Button "Polygon"<br>Set the Show/Hide toggle to "False"<br>**Expected: The Map should not show the Polygon with locations = ['Areeiro Lisbon, PT', 'Alfama Lisbon, PT', 'Jardim da Estrela, Lisbon, PT']. The Map should only have 1 polyline.**
Platform “if” true Polygon | Click on the Button "Polygon"<br>Set the Show/Hide toggle to "True"<br>**Expected: The Map should not show the Polygon with locations = ['Areeiro Lisbon, PT', 'Alfama Lisbon, PT', 'Jardim da Estrela, Lisbon, PT']. The Map should have 2 polylines.**
Polygon AllowDrag | Click on the Button "Polygon"<br>Set the AllowDrag toggle to "False"<br>Drag and Drop on the Polygon <br>**Expected: The Polygon should not be draggable**
Polygon don't AllowDrag | Click on the Button "Polygon"<br>Set the AllowDrag toggle to "True"<br>Drag and Drop on the Polygon <br>**Expected: The Polygon should be draggable**
Polygon AllowEdit | Click on the Button "Polygon"Set the AllowEdit toggle to "True"<br>Check if the Polygon has clickable/draggable dots all over the line.<br>**Expected: The Polygon should be editable**
Polygon don't AllowEdit | Click on the Button "Polygon"<br>Set the AllowEdit toggle to "False"<br>Check if the Polygon has clickable/draggable dots all over the line.<br>**Expected: The Polygon should not be editable**
Polygon StrokeColor by text | Click on the Button "Polygon"<br>Set the color input to "Red"<br>**Expected: The Polygon should be Red**
Polygon StrokeColor by hex | Click on the Button "Polygon"<br>Set the color input to "#FFFF00"<br>**Expected: The Polygon should be Yellow**
Polygon StrokeOpacity with transparency | Click on the Button "Polygon"<br>Set the opacity input to 0.2<br>**Expected: The Polygon should have some transparency**
Polygon StrokeOpacity Opaque | Click on the Button "Polygon"<br>Set the opacity input to 0.2<br>Set the opacity input to 1<br>**Expected: The Polygon should not have transparency**
Polygon StrokeWeight larger | Click on the Button "Polygon"<br>Set the weight input to 15<br>**Expected: The Polygon should have a larger line**
Polygon StrokeWeight slimer | Click on the Button "Polygon"<br>Set the weight input to 15<br>Set the weight input to 1<br>**Expected: The Polygon should have a slimer line**
Polygon OnClick | Click on the Button "Polygon"<br>Click on the Polygon with locations = "['Areeiro Lisbon, PT', 'Alfama Lisbon, PT', 'Jardim da Estrela, Lisbon, PT']" (black line)<br>**Expected: The sample has a feedback message "You triggered this method by clicking on the Shape Polyshape1 (Map: MyMap)"**
Polygon FillColor | Click on the Button "Polygon"<br>Set the color input to "Red"<br>**Expected: The Polygon should be Red**
Polygon FillOpacity with transparency | Click on the Button "Polygon"<br>Set the opacity input to 0.<br>**Expected: The Polygon should have some transparency**
Polygon FillOpacity Opaque | Click on the Button "Polygon"<br>Set the opacity input to 0.2<br>Set the opacity input to 1<br>**Expected: The Polygon should not have transparency**
CFG_InvalidPolygonShapeLocations - 2 locations | Click on the Button "Polygon"<br>Set the Locations textarea with ['Areeiro Lisbon, PT']<br>**Expected: The sample should show a feedback message with the error "The Polyline needs at least 2 valid address/coordinates".**


</body>
</html>


### Screenshots
![image](https://user-images.githubusercontent.com/6432232/128385962-532abdb2-56fd-4242-b8df-97c401e2599f.png)
![image](https://user-images.githubusercontent.com/6432232/128386003-e5a4ab6b-5e0e-45f0-8eaf-89b5625ea74e.png)


### Checklist
* [x] requires changes in OutSystems (if so, provide a module with changes) - **RGRIDT-858**
* [x] requires new sample page in OutSystems (if so, provide a module with changes) - done :)
